### PR TITLE
[spmd_types] fix MoE routing bias TP reduce

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #4160
* #4085
* __->__ #4154

maybe I missed this, caught while trying to migrate backends, MoE training w/ TP wasn't bitwise, as spmd_types didn't reduce routing bias over TP axis. Refactor logic to get local tensor, reduce TP, then reduce DP+CP. Set test to spmd_types backend without updating 100-step loss reference.

`NGPU=2 ./run_train.sh --module qwen3 --config qwen3_moe_debug --parallelism.tensor_parallel_degree 2 --parallelism.spmd_backend spmd_types --training.disable_cuda_graphs --training.steps 10 --debug.seed 42 --debug.deterministic`

### CODE REVIEW

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -480,16 +481,22 @@ def _update_expert_bias(
...
+                tokens_per_expert_E_by_layer = tokens_per_expert_E_by_layer.to_local()
+                tp_group = dtensor_mesh.get_group("tp")
+            elif parallel_dims.tp > 1:
+                tp_group = spmd_dense_mesh()["tp"].get_group()
+            else:
+                tp_group = None
+            if tp_group is not None:
+                torch.distributed.all_reduce(
+                    tokens_per_expert_E_by_layer,
+                    group=tp_group,
```
Comment (tianyu-l) [inline torchtitan/components/optimizer.py]: I think it might not have been your miss.
@acisseJZhong and I recently changed shardings of router module, see latest https://github.com/pytorch/torchtitan/blob/main/assets/images/moe_sharding.png

Right now when EP > 1, we need this reduction; when EP==1, we don't. Can we make the code here conditional?

**Review by tianyu-l — APPROVED** [review]
stamp to unblock, but please make sure we get correct result for [enable EP, disable EP] x [enable SP, disable SP] when TP > 1.

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/optimizer.py]: so my claim is that when EP not used, here you'll get a `TP` multiplier in the result -- but I guess the numerics won't change because it has no real effect

File: torchtitan/components/optimizer.py
Comment (pianpwk) [inline torchtitan/components/optimizer.py:466]: annoyingly, this is causing implicit P->R redistribute in AC case, gonna restructure this block

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (pianpwk) [reply torchtitan/components/optimizer.py]: ah yes thanks, gating on EP too is the right call

File: torchtitan/components/optimizer.py
Comment (pianpwk) [reply torchtitan/components/optimizer.py:466]: I think this new flow is cleaner:
- get local expert count tensor
- normalize by AC factor, store local
- stack counts across layers
- TP reduce, loss mesh reduce
- rewrap as DTensor if needed
